### DISCUSSION TITLE
Compatibility of lora and radix attention.

### DISCUSSION THREAD

**Discussion by walker-ai** [sgl-project/sglang#2141] [Q&A] [answered]
I tried to use LoRA Linear from sglang to implement a feature in my own code which is about inferencing with LoRA, and I have already incorporated the RadixAttn. But I found that the output is always `<eos>` which means nothing. For example:

```python
# prompt: "who are you"
# answer: ""
```

Until I saw the 

```python
assert (
    self.max_loras_per_batch > 0
    # FIXME
    and (self.lora_paths is None or self.disable_cuda_graph)
    and (self.lora_paths is None or self.disable_radix_cache)
), "compatibility of lora and cuda graph and radix attention is in progress"
assert self.base_gpu_id >= 0, "base_gpu_id must be non-negative"
```

I'm afraid that the question I met is caused by this? I m not sure about that, could you give me some advice? 

@Ying1123 Thx for your time!

**Comment by Galaxy-Husky** [comment]
I think you should disable `cuda_graph` and `radix_cache` when loading lora adapters like:
https://github.com/sgl-project/sglang/blob/f8b0326934bacb7a7d4eba68fb6eddebaa6ff751/test/srt/models/test_lora.py#L85-L94

**Reply by walker-ai** [reply]
Thank you for your reply! I noticed the update about apply S-LoRA method to adapt multi-lora switch, the code comments says that _**Not compatible with RadixCache currently**_. I wanna know if this is fundamentally infeasible, or since the intermediate computation results generated by different LoRA adapters are all different, making it difficult to reuse KV between requests (similar to Radix Cache), is this technically challenging to implement?

**Reply by Galaxy-Husky** [reply]
> Thank you for your reply! I noticed the update about apply S-LoRA method to adapt multi-lora switch, the code comments says that _**Not compatible with RadixCache currently**_. I wanna know if this is fundamentally infeasible, or since the intermediate computation results generated by different LoRA adapters are all different, making it difficult to reuse KV between requests (similar to Radix Cache), is this technically challenging to implement?

Sorry, I'm not a developer and I'm not sure about your question. You might get a faster official response by submitting an issue. Good luck!

**Reply by walker-ai** [reply]
thx anyway, I'll ask in SGL slack

**Reply by Kayce001** [reply]
i meet the same problem,does there is a solution?

**Reply by lifuhuang** [reply]
@Kayce001 , Radix support has now been merged in #7216 .

**Comment by lifuhuang** [comment, marked as answer]
**Update (8/22): it's already supported in #7216 by @Fridge003 . Please give it a try and let us know if you run into any issues.**

------------------

Hi, we are currently working on supporting compatibility between RadixCache and LoRA (cc @Fridge003 ). It's not a trivial feature because we essentially needs to keep separate KV cache per LoRA.

You can refer to #2929 for all LoRA-related planned work for H2.
